sandbox/apparmor/profile.go

```diff
@@ -251,3 +286,68 @@ var SnapConfineDistroProfilePath = func() string {
 
 	return ""
 }
+
+// SetupSnapConfineSnippets inspects the system and sets up local apparmor
+// policy for snap-confine. Local policy is included by the system-wide policy.
+// Returns whether any modifications was made to the snap-confine snippets.
+func SetupSnapConfineSnippets() (wasChanged bool, err error) {
+	// Create the local policy directory if it is not there.
+	if err := os.MkdirAll(dirs.SnapConfineAppArmorDir, 0755); err != nil {
+		return false, fmt.Errorf("cannot create snap-confine policy directory: %s", err)
+	}
+
+	// Location of the generated policy.
+	glob := "*"
+	policy := make(map[string]osutil.FileState)
+
+	// Check if NFS is mounted at or under $HOME. Because NFS is not
+	// transparent to apparmor we must alter our profile to counter that and
+	// allow snap-confine to work.
+	if nfs, err := osutilIsHomeUsingNFS(); err != nil {
+		logger.Noticef("cannot determine if NFS is in use: %v", err)
+	} else if nfs {
+		policy["nfs-support"] = &osutil.MemoryFileState{
+			Content: []byte(NfsSnippet),
+			Mode:    0644,
+		}
+		logger.Noticef("snapd enabled NFS support, additional implicit network permissions granted")
+	}
+
+	// Check if '/' is on overlayfs. If so, add the necessary rules for
+	// upperdir and allow snap-confine to work.
+	if overlayRoot, err := osutilIsRootWritableOverlay(); err != nil {
+		logger.Noticef("cannot determine if root filesystem on overlay: %v", err)
+	} else if overlayRoot != "" {
+		snippet := strings.Replace(OverlayRootSnippet, "###UPPERDIR###", overlayRoot, -1)
+		policy["overlay-root"] = &osutil.MemoryFileState{
+			Content: []byte(snippet),
+			Mode:    0644,
+		}
+		logger.Noticef("snapd enabled root filesystem on overlay support, additional upperdir permissions granted")
+	}
+
+	// Check whether apparmor_parser supports bpf capability. Some older
+	// versions do not, hence the capability cannot be part of the default
+	// profile of snap-confine as loading it would fail.
+	if features, err := ParserFeatures(); err != nil {
+		logger.Noticef("cannot determine apparmor_parser features: %v", err)
+	} else if strutil.ListContains(features, "cap-bpf") {
+		policy["cap-bpf"] = &osutil.MemoryFileState{
+			Content: []byte(capabilityBPFSnippet),
+			Mode:    0644,
+		}
+	}
+
+	// Ensure that generated policy is what we computed above.
+	created, removed, err := osutil.EnsureDirState(dirs.SnapConfineAppArmorDir, glob, policy)
+	if err != nil {
+		return false, fmt.Errorf("cannot synchronize snap-confine policy: %s", err)
+	}
+	return len(created) > 0 || len(removed) > 0, nil
+}
+
+func RemoveSnapConfineSnippets() error {
```

Comment:
this should have a doc comment and could use a unit test in the package

Reply:
Fixed!